keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -0,0 +1,229 @@
+* [Storage migration triggering controller](#storage-migration-triggering-controller)
+* [Implications to cluster operators](#implications-to-cluster-operators)
+* [Lifecycle of a StorageState object](#lifecycle-of-a-storagestate-object)
+* [Future work: HA clusters](#future-work-ha-clusters)
+* [Future work: persisted discovery document](#future-work-persisted-discovery-document)
+
+## Goal
+
+As the discovery document now exposes the [storage version hash][], storage version
+migration should start automatically when the hash changes.
+
+[storage version hash]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md
+
+## API design
+
+We introduce the `StorageState` API, as a CRD defined in the `migration.k8s.io`
+group.
+
+```golang
+// StorageState is the state of the storage, for a specific resource.
+type StorageState struct {
+  metav1.TypeMeta
+  // The name is "<resource>.<group>". The API validation will enforce it by
+  // comapring the name with the spec.resource.
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+type StorageStateSpec {
+  // The resource this StorageState is about.
+  Resource GroupResource
+}
+
+type StorageStateStatus {
+  // The hash values of possible stored versions, the versions to which the
+  // persisted instances of spec.resource are encoded.
+  // "Unknown" is a valid value in the list, and is the default value.
+  // It is not safe to upgrade or downgrade to an apiserver binary that does not
+  // support all versions listed in this field, or if "Unknown" is listed.
+  // Once the storage version migration for this resource has completed, the
+  // value of this field is refined to only contain the storageVersionHash.
+  // Once the apiserver has changed the storage version, the new storage version
+  // is appended to the list.
+  // +optional
+  StoredVersionsHashes []string
+  // The hash value of the storage version, as shown in the discovery document
+  // served by the API server.
+  // Storage Version is the version to which objects are converted to
+  // before persisted.
+  StorageVersionHash string
+  // LastHeartbeatTime is the last time the storage migration triggering
+  // controller checks the storage version hash of this resource in the
+  // discovery document and updates this field.
+  // +optional
+  LastHeartbeatTime metav1.Time
+}
+```
+
+We had considered making `StoredVersionsHashes` part of the `spec` instead of
+the `status`, because the stored versions cannot be deduced immediately by
+inspecting Kubernetes API. However, we decided to put it in the status because
+* Its value eventually is determined by the `StorageVersionMigration` API, i.e.,
+  it's set to the storage version when the corresponding migration has
+  completed.
+* Even though we cannot reconstruct this field immediately, this field is
+  initialized to empty, meaning "unknown", which truthfully reflects the state.
```

Comment:
"unknown" helps a lot, thanks.

Reply:
i guess you mean initialized to ["Unknown"] instead of 'empty, meaning "known"'